server/types/domain.go

```diff
@@ -0,0 +1,234 @@
+func (d DomainType) IsUnbounded() bool {
+	return d.AsType.IsUnbounded()
+}
+
+// MaxSerializedWidth implements the types.ExtendedType interface.
+func (d DomainType) MaxSerializedWidth() types.ExtendedTypeSerializedWidth {
+	return d.AsType.MaxSerializedWidth()
+}
+
+// MaxTextResponseByteLength implements the DoltgresType interface.
+func (d DomainType) MaxTextResponseByteLength(ctx *sql.Context) uint32 {
+	return d.AsType.MaxTextResponseByteLength(ctx)
+}
+
+// OID implements the DoltgresType interface.
+func (d DomainType) OID() uint32 {
+	//TODO: generate unique oid
+	return d.AsType.OID()
+}
+
+// Promote implements the DoltgresType interface.
+func (d DomainType) Promote() sql.Type {
+	return d.AsType.Promote()
+}
+
+// SerializedCompare implements the DoltgresType interface.
+func (d DomainType) SerializedCompare(v1 []byte, v2 []byte) (int, error) {
+	return d.AsType.SerializedCompare(v1, v2)
+}
+
+// SQL implements the DoltgresType interface.
+func (d DomainType) SQL(ctx *sql.Context, dest []byte, v interface{}) (sqltypes.Value, error) {
+	return d.AsType.SQL(ctx, dest, v)
+}
+
+// String implements the DoltgresType interface.
+func (d DomainType) String() string {
+	return d.Name
+}
+
+// ToArrayType implements the DoltgresType interface.
+func (d DomainType) ToArrayType() DoltgresArrayType {
+	return d.AsType.ToArrayType()
+}
+
+// Type implements the DoltgresType interface.
+func (d DomainType) Type() query.Type {
+	return d.AsType.Type()
+}
+
+// ValueType implements the DoltgresType interface.
+func (d DomainType) ValueType() reflect.Type {
+	return d.AsType.ValueType()
+}
+
+// Zero implements the DoltgresType interface.
+func (d DomainType) Zero() interface{} {
+	return d.AsType.Zero()
+}
+
+// SerializeType implements the types.ExtendedType interface.
+func (d DomainType) SerializeType() ([]byte, error) {
+	b := SerializationId_Domain.ToByteSlice(0)
+	writer := utils.NewWriter(256)
+	writer.String(d.Schema)
+	writer.String(d.Name)
+	writer.String(d.DefaultExpr)
+	writer.Bool(d.NotNull)
+	writer.VariableUint(uint64(len(d.Checks)))
+	for _, check := range d.Checks {
+		writer.String(check.Name)
+		writer.String(check.CheckExpression)
+	}
+	asTyp, err := d.AsType.SerializeType()
+	if err != nil {
+		return nil, err
+	}
+	b = append(b, writer.Data()...)
+	return append(b, asTyp...), nil
+}
+
+func (d DomainType) deserializeType(version uint16, metadata []byte) (DoltgresType, error) {
+	switch version {
+	case 0:
+		reader := utils.NewReader(metadata)
+		d.Schema = reader.String()
+		d.Name = reader.String()
+		d.DefaultExpr = reader.String()
+		d.NotNull = reader.Bool()
+		numOfChecks := reader.VariableUint()
+		for k := uint64(0); k < numOfChecks; k++ {
+			checkName := reader.String()
+			checkExpr := reader.String()
+			d.Checks = append(d.Checks, &sql.CheckDefinition{
+				Name:            checkName,
+				CheckExpression: checkExpr,
+				Enforced:        true,
+			})
+		}
+		t, err := DeserializeType(reader.RemainingBytes())
```

Comment:
The preferred way of doing this would be `metadata[len(metadata)-reader.RemainingBytes():]`, which gets us the same result as the changed `reader.RemainingBytes`.

We could also add `reader.BytesRead()` or `reader.CurrentOffset()` to make it easier to get how much we've read. Then it'd be `metadata[reader.BytesRead():]`. Definitely don't want to add a way to arbitrarily read data from the reader though.